// @vkontakte/vkjs/lib/htmlEntities.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.decodeHTMLEntities = decodeHTMLEntities;
exports.decodeHTMLEntitiesDeep = decodeHTMLEntitiesDeep;
exports.encodeHTMLEntities = encodeHTMLEntities;
exports.escape = escape;
exports.outOfBoundsChar = void 0;
exports.unescape = unescape;
var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _codepoints = require("./lib/codepoints");
/* eslint-disable @typescript-eslint/quotes */

var escapeMap = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  "'": '&#39;',
  '"': '&quot;'
};
var unescapeMap = {
  '&amp;': '&',
  '&#38;': '&',
  '&lt;': '<',
  '&#60;': '<',
  '&gt;': '>',
  '&#62;': '>',
  '&apos;': "'",
  '&#39;': "'",
  '&quot;': '"',
  '&#34;': '"'
};
var namedEntities = [['&amp;', '&'], ['&lt;', '<'], ['&gt;', '>'], ['&quot;', '"'], ['&apos;', "'"]];
var ESCAPE_REGEX = /[&<>'"]/g;

/**
 * Safely escape HTML entities such as `&`, `<`, `>`, `"`, and `'`
 * @param {string} input
 */
function escape(input) {
  if (input == null) {
    return '';
  }
  return input.replace(ESCAPE_REGEX, function (entity) {
    return escapeMap[entity];
  });
}
var UNESCAPE_REGEX = /&(?:amp|#38|lt|#60|gt|#62|apos|#39|quot|#34);/g;

/**
 * Unescape HTML entities such as `&`, `<`, `>`, `"`, and `'`
 * @param {string} input
 */
function unescape(input) {
  if (input == null) {
    return '';
  }
  return input.replace(UNESCAPE_REGEX, function (entity) {
    return unescapeMap[entity];
  });
}
var outOfBoundsChar = String.fromCharCode(65533);
exports.outOfBoundsChar = outOfBoundsChar;
var ENCODE_REGEX = /(?:[<>'"&\x01-\x08\x11-\x15\x17-\x1F\x7f-\uD7FF\uE000-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])/g;
function encodeHTMLEntities(input) {
  if (input == null) {
    return '';
  }
  return input.replace(ENCODE_REGEX, function (entity) {
    var code = entity.length > 1 ? (0, _codepoints.getCodePointAt)(entity, 0) : entity.charCodeAt(0);
    return '&#' + String(code) + ';';
  });
}
var DECODE_REGEX = /&(?:#\d+|#[xX][\da-fA-F]+|[0-9a-zA-Z]+);/g;
function decodeHTMLEntities(input) {
  if (input == null) {
    return '';
  }
  input = namedEntities.reduce(function (result, _ref) {
    var _ref2 = (0, _slicedToArray2["default"])(_ref, 2),
      mask = _ref2[0],
      _char = _ref2[1];
    return result.replace(new RegExp(mask, 'ig'), _char);
  }, input);
  return input.replace(DECODE_REGEX, function (entity) {
    if (entity[0] === '&' && entity[1] === '#') {
      var secondChar = entity[2];
      var code = secondChar === 'x' || secondChar === 'X' ? parseInt(entity.substr(3), 16) : parseInt(entity.substr(2));
      if (code >= 0x10ffff) {
        return outOfBoundsChar;
      }
      if (code > 65535) {
        return (0, _codepoints.fromCodePoint)(code);
      }
      return String.fromCharCode(_codepoints.numericUnicodeMap[code] || code);
    }
    return entity;
  });
}
function decodeHTMLEntitiesDeep(input) {
  if (typeof input === 'string') {
    return decodeHTMLEntities(input);
  }
  if ((0, _typeof2["default"])(input) === 'object') {
    var correctType = Object.prototype.toString.apply(input);
    if (correctType === '[object Array]') {
      return input.map(function (item) {
        return decodeHTMLEntitiesDeep(item);
      });
    }
    if (correctType === '[object Object]') {
      var response = {};
      Object.keys(input).forEach(function (item) {
        response[decodeHTMLEntities(item)] = decodeHTMLEntitiesDeep(input[item]);
      });
      return response;
    }
  }
  return input;
}
